TestLeasingReconnectNonOwnerGet fails

```
=== RUN   TestLeasingReconnectNonOwnerGet
2017-08-06 23:02:57.425971 I | integration: launching 2403088238183459612 (unix://localhost:24030882381834596120)
2017-08-06 23:02:57.428075 I | etcdserver: name = 2403088238183459612
2017-08-06 23:02:57.428086 I | etcdserver: data dir = /tmp/etcd878134323
2017-08-06 23:02:57.428093 I | etcdserver: member dir = /tmp/etcd878134323/member
2017-08-06 23:02:57.428100 I | etcdserver: heartbeat = 10ms
2017-08-06 23:02:57.428106 I | etcdserver: election = 100ms
2017-08-06 23:02:57.428113 I | etcdserver: snapshot count = 0
2017-08-06 23:02:57.428122 I | etcdserver: advertise client URLs = unix://127.0.0.1:2129632169
2017-08-06 23:02:57.428129 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:2129532169
2017-08-06 23:02:57.428139 I | etcdserver: initial cluster = 2403088238183459612=unix://127.0.0.1:2129532169
2017-08-06 23:02:57.433322 I | etcdserver: starting member 862b8cf933b105a5 in cluster 781199ec5e4f472d
2017-08-06 23:02:57.433351 I | raft: 862b8cf933b105a5 became follower at term 0
2017-08-06 23:02:57.433367 I | raft: newRaft 862b8cf933b105a5 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2017-08-06 23:02:57.433377 I | raft: 862b8cf933b105a5 became follower at term 1
2017-08-06 23:02:57.435645 W | auth: simple token is not cryptographically signed
2017-08-06 23:02:57.437269 I | etcdserver: set snapshot count to default 100000
2017-08-06 23:02:57.437281 I | etcdserver: starting server... [version: 3.2.0+git, cluster version: to_be_decided]
2017-08-06 23:02:57.437534 I | integration: launched 2403088238183459612 (unix://localhost:24030882381834596120)
2017-08-06 23:02:57.438619 I | etcdserver/membership: added member 862b8cf933b105a5 [unix://127.0.0.1:2129532169] to cluster 781199ec5e4f472d
2017-08-06 23:02:57.523606 I | raft: 862b8cf933b105a5 is starting a new election at term 1
2017-08-06 23:02:57.523630 I | raft: 862b8cf933b105a5 became candidate at term 2
2017-08-06 23:02:57.523644 I | raft: 862b8cf933b105a5 received MsgVoteResp from 862b8cf933b105a5 at term 2
2017-08-06 23:02:57.523660 I | raft: 862b8cf933b105a5 became leader at term 2
2017-08-06 23:02:57.523671 I | raft: raft.node: 862b8cf933b105a5 elected leader 862b8cf933b105a5 at term 2
2017-08-06 23:02:57.523971 I | etcdserver: setting up the initial cluster version to 3.2
2017-08-06 23:02:57.524092 I | etcdserver: published {Name:2403088238183459612 ClientURLs:[unix://127.0.0.1:2129632169]} to cluster 781199ec5e4f472d
2017-08-06 23:02:57.524272 N | etcdserver/membership: set the initial cluster version to 3.2
2017-08-06 23:02:57.538693 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken write unix @->localhost:24030882381834596120: write: broken pipe.
2017-08-06 23:02:57.539992 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.541117 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.542278 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.543420 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.544540 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.545723 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.546869 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.547949 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017-08-06 23:02:57.556464 I | etcdserver/api/v3rpc: transport: http2Client.notifyError got notified that the client transport was broken write unix @->localhost:24030882381834596120: write: broken pipe.
2017-08-06 23:02:57.556531 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing context canceled"; Reconnecting to {localhost:24030882381834596120 <nil>}
2017-08-06 23:02:57.556598 I | integration: terminating 2403088238183459612 (unix://localhost:24030882381834596120)
2017-08-06 23:02:57.556670 I | etcdserver/api/v3rpc: get error from resetTransport context canceled, transportMonitor returning
2017-08-06 23:02:57.556678 I | etcdserver/api/v3rpc: grpc: addrConn.transportMonitor exits due to: context canceled
2017-08-06 23:02:57.560413 I | integration: terminated 2403088238183459612 (unix://localhost:24030882381834596120)
--- FAIL: TestLeasingReconnectNonOwnerGet (0.19s)
	leasing_test.go:1612: rpc error: code = Unavailable desc = transport: write unix @->localhost:24030882381834596120: write: broken pipe
```

https://jenkins-etcd-public.prod.coreos.systems/job/etcd-ci-ppc64/1820/consoleFull

--- a/clientv3/integration/leasing_test.go
+++ b/clientv3/integration/leasing_test.go
@@ -1296,7 +1296,7 @@ func TestLeasingReconnectOwnerRevoke(t *testing.T) {
 		defer close(sdonec)
 		for i := 0; i < 10 && cctx.Err() == nil; i++ {
 			clus.Members[0].Stop(t)
-			time.Sleep(100 * time.Millisecond)
+			time.Sleep(10 * time.Millisecond)
 			clus.Members[0].Restart(t)
 		}
 	}()
@@ -1317,7 +1317,7 @@ func TestLeasingReconnectOwnerRevoke(t *testing.T) {
 	case <-pdonec:
 		cancel()
 		<-sdonec
-	case <-time.After(5 * time.Second):
+	case <-time.After(10 * time.Second):
 		cancel()
 		<-sdonec
 		<-pdonec

--- a/clientv3/leasing/kv.go
+++ b/clientv3/leasing/kv.go
@@ -20,9 +20,12 @@ import (
 
 	v3 "github.com/coreos/etcd/clientv3"
 	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
 	"github.com/coreos/etcd/mvcc/mvccpb"
 
 	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
 )
 
 type leasingKV struct {
@@ -239,16 +242,33 @@ func (lkv *leasingKV) put(ctx context.Context, op v3.Op) (pr *v3.PutResponse, er
 }
 
 func (lkv *leasingKV) acquire(ctx context.Context, key string, op v3.Op) (*v3.TxnResponse, error) {
-	if err := lkv.waitSession(ctx); err != nil {
-		return nil, err
+	for ctx.Err() == nil {
+		if err := lkv.waitSession(ctx); err != nil {
+			return nil, err
+		}
+		resp, err := lkv.kv.Txn(ctx).If(
+			v3.Compare(v3.CreateRevision(lkv.pfx+key), "=", 0)).
+			Then(
+				op,
+				v3.OpPut(lkv.pfx+key, "", v3.WithLease(lkv.leaseID()))).
+			Else(
+				op,
+				v3.OpGet(lkv.pfx+key),
+			).Commit()
+		if err == nil {
+			if !resp.Succeeded {
+				kvs := resp.Responses[1].GetResponseRange().Kvs
+				// if txn failed since already owner, lease is acquired
+				resp.Succeeded = v3.LeaseID(kvs[0].Lease) == lkv.leaseID()
+			}
+			return resp, nil
+		}
+		// retry if transient error
+		if _, ok := err.(rpctypes.EtcdError); ok || grpc.Code(err) != codes.Unavailable {
+			return nil, err
+		}
 	}
-	return lkv.kv.Txn(ctx).If(
-		v3.Compare(v3.CreateRevision(lkv.pfx+key), "=", 0)).
-		Then(
-			op,
-			v3.OpPut(lkv.pfx+key, "", v3.WithLease(lkv.leaseID()))).
-		Else(op).
-		Commit()
+	return nil, ctx.Err()
 }
 
 func (lkv *leasingKV) get(ctx context.Context, op v3.Op) (*v3.GetResponse, error) {

